File: prototypes/b01-architecture-evidence/adr-example-semantics/rust/src/strict_json.rs
use std::collections::BTreeMap;

use serde_json::{Map, Number, Value};

use crate::error::{EngineError, EngineResult};

const MAX_LANGUAGE_NEUTRAL_INTEGER: u64 = 9_007_199_254_740_991;

#[derive(Clone, Copy, Debug)]
pub(crate) struct JsonLimits {
    pub(crate) maximum_input_bytes: usize,
    pub(crate) maximum_json_depth: usize,
    pub(crate) maximum_json_nodes: usize,
    pub(crate) maximum_object_members: usize,
    pub(crate) maximum_array_items: usize,
    pub(crate) maximum_key_utf8_bytes: usize,
    pub(crate) maximum_string_utf8_bytes: usize,
    pub(crate) maximum_total_string_utf8_bytes: usize,
    pub(crate) maximum_integer_characters: usize,
    pub(crate) allow_floats: bool,
}

impl JsonLimits {
    pub(crate) const fn corpus_bootstrap() -> Self {
        Self {
            maximum_input_bytes: 262_144,
            maximum_json_depth: 32,
            maximum_json_nodes: 100_000,
            maximum_object_members: 4_096,
            maximum_array_items: 4_096,
            maximum_key_utf8_bytes: 256,
            maximum_string_utf8_bytes: 65_536,
            maximum_total_string_utf8_bytes: 131_072,
            maximum_integer_characters: 32,
            allow_floats: false,
        }
    }
}

pub(crate) fn parse_strict(input: &[u8], limits: JsonLimits) -> EngineResult<Value> {
    if input.is_empty() {
        return Err(EngineError::json("input is empty"));
    }
    if input.len() > limits.maximum_input_bytes {
        return Err(EngineError::json(format!(
            "input has {} bytes; limit is {}",
            input.len(),
            limits.maximum_input_bytes
        )));
    }
    if input.starts_with(&[0xef, 0xbb, 0xbf]) {
        return Err(EngineError::json("UTF-8 BOM is forbidden"));
    }
    std::str::from_utf8(input)
        .map_err(|error| EngineError::json(format!("input is not UTF-8: {error}")))?;
    if limits.allow_floats {
        return Err(EngineError::json(
            "this engine's closed profile requires allow_floats=false",
        ));
    }

    let mut parser = Parser {
        input,
        position: 0,
        limits,
        nodes: 0,
        total_string_bytes: 0,
    };
    parser.skip_whitespace();
    let value = parser.parse_value(1)?;
    parser.skip_whitespace();
    if parser.position != input.len() {
        return Err(parser.error("trailing non-whitespace content"));
    }
    Ok(value)
}

struct Parser<'input> {
    input: &'input [u8],
    position: usize,
    limits: JsonLimits,
    nodes: usize,
    total_string_bytes: usize,
}

impl Parser<'_> {
    fn parse_value(&mut self, depth: usize) -> EngineResult<Value> {
        if depth > self.limits.maximum_json_depth {
            return Err(self.error("maximum JSON depth exceeded"));
        }
        self.nodes = self
            .nodes
            .checked_add(1)
            .ok_or_else(|| self.error("JSON node count overflow"))?;
        if self.nodes > self.limits.maximum_json_nodes {
            return Err(self.error("maximum JSON node count exceeded"));
        }

        match self.peek() {
            Some(b'{') => self.parse_object(depth),
            Some(b'[') => self.parse_array(depth),
            Some(b'"') => self.parse_string(false).map(Value::String),
            Some(b't') => self.parse_literal(b"true", Value::Bool(true)),
            Some(b'f') => self.parse_literal(b"false", Value::Bool(false)),
            Some(b'n') => self.parse_literal(b"null", Value::Null),
            Some(b'-' | b'0'..=b'9') => self.parse_integer(),
            Some(_) => Err(self.error("invalid JSON value")),
            None => Err(self.error("unexpected end of input")),
        }
    }

    fn parse_object(&mut self, depth: usize) -> EngineResult<Value> {
        self.consume_exact(b'{')?;
        self.skip_whitespace();
        let mut object = Map::new();
        if self.consume_if(b'}') {
            return Ok(Value::Object(object));
        }

        loop {
            if object.len() >= self.limits.maximum_object_members {
                return Err(self.error("maximum object member count exceeded"));
            }
            if self.peek() != Some(b'"') {
                return Err(self.error("object key must be a string"));
            }
            let key = self.parse_string(true)?;
            if object.contains_key(&key) {
                return Err(self.error(format!("duplicate object key {key:?}")));
            }
            self.skip_whitespace();
            self.consume_exact(b':')?;
            self.skip_whitespace();
            let value = self.parse_value(depth.saturating_add(1))?;
            object.insert(key, value);
            self.skip_whitespace();
            if self.consume_if(b'}') {
                return Ok(Value::Object(object));
            }
            self.consume_exact(b',')?;
            self.skip_whitespace();
        }
    }

    fn parse_array(&mut self, depth: usize) -> EngineResult<Value> {
        self.consume_exact(b'[')?;
        self.skip_whitespace();
        let mut array = Vec::new();
        if self.consume_if(b']') {
            return Ok(Value::Array(array));
        }

        loop {
            if array.len() >= self.limits.maximum_array_items {
                return Err(self.error("maximum array item count exceeded"));
            }
            array.push(self.parse_value(depth.saturating_add(1))?);
            self.skip_whitespace();
            if self.consume_if(b']') {
                return Ok(Value::Array(array));
            }
            self.consume_exact(b',')?;
            self.skip_whitespace();
        }
    }

    fn parse_string(&mut self, is_key: bool) -> EngineResult<String> {
        self.consume_exact(b'"')?;
        let mut decoded = Vec::new();

        loop {
            let byte = self
                .next()
                .ok_or_else(|| self.error("unterminated JSON string"))?;
            match byte {
                b'"' => break,
                b'\\' => self.parse_escape(&mut decoded)?,
                0x00..=0x1f => {
                    return Err(self.error("unescaped control character in JSON string"));
                }
                _ => decoded.push(byte),
            }
        }

        let value = String::from_utf8(decoded)
            .map_err(|error| self.error(format!("invalid UTF-8 string: {error}")))?;
        let individual_limit = if is_key {
            self.limits.maximum_key_utf8_bytes
        } else {
            self.limits.maximum_string_utf8_bytes
        };
        if value.len() > individual_limit {
            return Err(self.error(if is_key {
                "maximum key UTF-8 byte count exceeded"
            } else {
                "maximum string UTF-8 byte count exceeded"
            }));
        }
        self.total_string_bytes = self
            .total_string_bytes
            .checked_add(value.len())
            .ok_or_else(|| self.error("total string UTF-8 byte count overflow"))?;
        if self.total_string_bytes > self.limits.maximum_total_string_utf8_bytes {
            return Err(self.error("maximum total string UTF-8 byte count exceeded"));
        }
        Ok(value)
    }

    fn parse_escape(&mut self, decoded: &mut Vec<u8>) -> EngineResult<()> {
        let escaped = self
            .next()
            .ok_or_else(|| self.error("truncated JSON escape"))?;
        match escaped {
            b'"' | b'\\' | b'/' => decoded.push(escaped),
            b'b' => decoded.push(0x08),
            b'f' => decoded.push(0x0c),
            b'n' => decoded.push(b'\n'),
            b'r' => decoded.push(b'\r'),
            b't' => decoded.push(b'\t'),
            b'u' => {
                let first = self.parse_hex_quad()?;
                let scalar = if (0xd800..=0xdbff).contains(&first) {
                    self.consume_exact(b'\\')?;
                    self.consume_exact(b'u')?;
                    let second = self.parse_hex_quad()?;
                    if !(0xdc00..=0xdfff).contains(&second) {
                        return Err(self.error("high surrogate is not followed by a low surrogate"));
                    }
                    0x1_0000 + ((u32::from(first) - 0xd800) << 10) + (u32::from(second) - 0xdc00)
                } else if (0xdc00..=0xdfff).contains(&first) {
                    return Err(self.error("unpaired low surrogate"));
                } else {
                    u32::from(first)
                };
                let character = char::from_u32(scalar)
                    .ok_or_else(|| self.error("invalid Unicode scalar value"))?;
                let mut buffer = [0_u8; 4];
                decoded.extend_from_slice(character.encode_utf8(&mut buffer).as_bytes());
            }
            _ => return Err(self.error("invalid JSON escape")),
        }
        Ok(())
    }

    fn parse_hex_quad(&mut self) -> EngineResult<u16> {
        let mut value = 0_u16;
        for _ in 0..4 {
            let byte = self
                .next()
                .ok_or_else(|| self.error("truncated Unicode escape"))?;
            let digit = match byte {
                b'0'..=b'9' => u16::from(byte - b'0'),
                b'a'..=b'f' => u16::from(byte - b'a') + 10,
                b'A'..=b'F' => u16::from(byte - b'A') + 10,
                _ => return Err(self.error("invalid hexadecimal digit in Unicode escape")),
            };
            value = (value << 4) | digit;
        }
        Ok(value)
    }

    fn parse_integer(&mut self) -> EngineResult<Value> {
        let start = self.position;
        let negative = self.consume_if(b'-');
        match self.peek() {
            Some(b'0') => {
                self.position += 1;
                if matches!(self.peek(), Some(b'0'..=b'9')) {
                    return Err(self.error("integer has a leading zero"));
                }
                if negative {
                    return Err(self.error("negative zero is forbidden"));
                }
            }
            Some(b'1'..=b'9') => {
                self.position += 1;
                while matches!(self.peek(), Some(b'0'..=b'9')) {
                    self.position += 1;
                }
            }
            _ => return Err(self.error("invalid integer")),
        }
        if matches!(self.peek(), Some(b'.' | b'e' | b'E')) {
            return Err(self.error("floating-point JSON numbers are forbidden"));
        }
        let token = &self.input[start..self.position];
        if token.len() > self.limits.maximum_integer_characters {
            return Err(self.error("maximum integer character count exceeded"));
        }
        let token_text = std::str::from_utf8(token)
            .map_err(|error| self.error(format!("invalid integer token: {error}")))?;
        if negative {
            let parsed = token_text
                .parse::<i64>()
                .map_err(|error| self.error(format!("integer is out of range: {error}")))?;
            if parsed.unsigned_abs() > MAX_LANGUAGE_NEUTRAL_INTEGER {
                return Err(self.error("integer exceeds the language-neutral exact range"));
            }
            Ok(Value::Number(Number::from(parsed)))
        } else {
            let parsed = token_text
                .parse::<u64>()
                .map_err(|error| self.error(format!("integer is out of range: {error}")))?;
            if parsed > MAX_LANGUAGE_NEUTRAL_INTEGER {
                return Err(self.error("integer exceeds the language-neutral exact range"));
            }
            Ok(Value::Number(Number::from(parsed)))
        }
    }

    fn parse_literal(&mut self, expected: &[u8], value: Value) -> EngineResult<Value> {
        let end = self
            .position
            .checked_add(expected.len())
            .ok_or_else(|| self.error("literal position overflow"))?;
        if self.input.get(self.position..end) != Some(expected) {
            return Err(self.error("invalid JSON literal"));
        }
        self.position = end;
        Ok(value)
    }

    fn skip_whitespace(&mut self) {
        while matches!(self.peek(), Some(b' ' | b'\n' | b'\r' | b'\t')) {
            self.position += 1;
        }
    }

    fn consume_exact(&mut self, expected: u8) -> EngineResult<()> {
        if self.consume_if(expected) {
            Ok(())
        } else {
            Err(self.error(format!("expected byte {:?}", char::from(expected))))
        }
    }

    fn consume_if(&mut self, expected: u8) -> bool {
        if self.peek() == Some(expected) {
            self.position += 1;
            true
        } else {
            false
        }
    }

    fn peek(&self) -> Option<u8> {
        self.input.get(self.position).copied()
    }

    fn next(&mut self) -> Option<u8> {
        let byte = self.peek()?;
        self.position += 1;
        Some(byte)
    }

    fn error(&self, detail: impl Into<String>) -> EngineError {
        EngineError::json(format!("byte {}: {}", self.position, detail.into()))
    }
}

pub(crate) fn canonicalize(value: &Value) -> Value {
    match value {
        Value::Array(items) => Value::Array(items.iter().map(canonicalize).collect()),
        Value::Object(object) => {
            let sorted = object
                .iter()
                .map(|(key, value)| (key.clone(), canonicalize(value)))
                .collect::<BTreeMap<_, _>>();
            let canonical = sorted.into_iter().collect::<Map<_, _>>();
            Value::Object(canonical)
        }
        other => other.clone(),
    }
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::{parse_strict, JsonLimits};

    fn limits() -> JsonLimits {
        JsonLimits::corpus_bootstrap()
    }

    #[test]
    fn parse_strict_should_decode_surrogate_pair_and_reject_duplicate_decoded_key() {
        let parsed = parse_strict(br#"{"emoji":"\uD83D\uDE00"}"#, limits());
        assert_eq!(parsed.ok(), Some(json!({"emoji": "😀"})));

        let duplicate = parse_strict(br#"{"a":1,"\u0061":2}"#, limits());
        assert!(duplicate.is_err());
    }

    #[test]
    fn parse_strict_should_reject_floats_negative_zero_and_large_integer() {
        assert!(parse_strict(b"1.0", limits()).is_err());
        assert!(parse_strict(b"-0", limits()).is_err());
        assert!(parse_strict(b"9007199254740992", limits()).is_err());
    }

    #[test]
    fn parse_strict_should_enforce_depth_nodes_members_and_strings() {
        let mut bounded = limits();
        bounded.maximum_json_depth = 2;
        assert!(parse_strict(br#"{"a":{"b":1}}"#, bounded).is_err());

        bounded = limits();
        bounded.maximum_json_nodes = 2;
        assert!(parse_strict(br#"[1,2]"#, bounded).is_err());

        bounded = limits();
        bounded.maximum_object_members = 1;
        assert!(parse_strict(br#"{"a":1,"b":2}"#, bounded).is_err());

        bounded = limits();
        bounded.maximum_string_utf8_bytes = 1;
        assert!(parse_strict(br#""ab""#, bounded).is_err());
    }
}

File: prototypes/b01-architecture-evidence/adr-example-semantics/typescript/src/strict-json.ts
export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

export interface JsonLimits {
  readonly maxBytes: number;
  readonly maxDepth: number;
  readonly maxNodes: number;
  readonly maxMembers: number;
  readonly maxArrayItems: number;
  readonly maxKeyBytes: number;
  readonly maxStringBytes: number;
  readonly maxTotalStringBytes: number;
  readonly maxIntegerCharacters: number;
}

export class StrictJsonError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "StrictJsonError";
  }
}

const encoder = new TextEncoder();

export function strictJsonParse(
  bytes: Uint8Array,
  limits: JsonLimits,
): JsonValue {
  validateLimits(limits);
  if (bytes.byteLength > limits.maxBytes) {
    throw new StrictJsonError(
      `JSON byte length ${bytes.byteLength} exceeds ${limits.maxBytes}`,
    );
  }
  if (bytes[0] === 0xef && bytes[1] === 0xbb && bytes[2] === 0xbf) {
    throw new StrictJsonError("UTF-8 BOM is forbidden");
  }

  let text: string;
  try {
    text = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch {
    throw new StrictJsonError("JSON is not valid UTF-8");
  }
  return new Parser(text, limits).parse();
}

function validateLimits(limits: JsonLimits): void {
  for (const [name, value] of Object.entries(limits)) {
    if (!Number.isSafeInteger(value) || value <= 0) {
      throw new StrictJsonError(`${name} must be a positive safe integer`);
    }
  }
}

class Parser {
  private index = 0;
  private nodes = 0;
  private totalStringBytes = 0;

  constructor(
    private readonly text: string,
    private readonly limits: JsonLimits,
  ) {}

  parse(): JsonValue {
    this.skipWhitespace();
    const value = this.parseValue(1);
    this.skipWhitespace();
    if (this.index !== this.text.length) {
      this.fail("trailing bytes after the JSON value");
    }
    return value;
  }

  private parseValue(depth: number): JsonValue {
    if (depth > this.limits.maxDepth) {
      this.fail(`JSON depth exceeds ${this.limits.maxDepth}`);
    }
    this.nodes += 1;
    if (this.nodes > this.limits.maxNodes) {
      this.fail(`JSON node count exceeds ${this.limits.maxNodes}`);
    }

    const token = this.text[this.index];
    switch (token) {
      case "{":
        return this.parseObject(depth);
      case "[":
        return this.parseArray(depth);
      case '"':
        return this.parseString(false);
      case "t":
        this.expectLiteral("true");
        return true;
      case "f":
        this.expectLiteral("false");
        return false;
      case "n":
        this.expectLiteral("null");
        return null;
      default:
        if (token === "-" || isDigit(token)) {
          return this.parseInteger();
        }
        this.fail("expected a JSON value");
    }
  }

  private parseObject(depth: number): { [key: string]: JsonValue } {
    this.index += 1;
    this.skipWhitespace();
    const result: { [key: string]: JsonValue } = Object.create(null) as {
      [key: string]: JsonValue;
    };
    const seen = new Set<string>();
    let memberCount = 0;
    if (this.consume("}")) {
      return result;
    }

    while (true) {
      if (memberCount >= this.limits.maxMembers) {
        this.fail(`JSON object member count exceeds ${this.limits.maxMembers}`);
      }
      if (this.text[this.index] !== '"') {
        this.fail("object member name must be a JSON string");
      }
      const key = this.parseString(true);
      if (seen.has(key)) {
        this.fail(`duplicate object member ${JSON.stringify(key)}`);
      }
      seen.add(key);
      memberCount += 1;
      this.skipWhitespace();
      if (!this.consume(":")) {
        this.fail("expected ':' after object member name");
      }
      this.skipWhitespace();
      result[key] = this.parseValue(depth + 1);
      this.skipWhitespace();
      if (this.consume("}")) {
        return result;
      }
      if (!this.consume(",")) {
        this.fail("expected ',' or '}' after object member value");
      }
      this.skipWhitespace();
    }
  }

  private parseArray(depth: number): JsonValue[] {
    this.index += 1;
    this.skipWhitespace();
    const result: JsonValue[] = [];
    if (this.consume("]")) {
      return result;
    }
    while (true) {
      if (result.length >= this.limits.maxArrayItems) {
        this.fail(`array item count exceeds ${this.limits.maxArrayItems}`);
      }
      result.push(this.parseValue(depth + 1));
      this.skipWhitespace();
      if (this.consume("]")) {
        return result;
      }
      if (!this.consume(",")) {
        this.fail("expected ',' or ']' after array element");
      }
      this.skipWhitespace();
    }
  }

  private parseString(isKey: boolean): string {
    this.index += 1;
    let result = "";
    let byteLength = 0;
    while (this.index < this.text.length) {
      const code = this.text.charCodeAt(this.index);
      if (code === 0x22) {
        this.index += 1;
        this.totalStringBytes += byteLength;
        return result;
      }
      if (code < 0x20) {
        this.fail("unescaped control character in JSON string");
      }
      if (code === 0x5c) {
        this.index += 1;
        const fragment = this.parseEscape();
        byteLength = this.checkedStringAppend(
          byteLength,
          utf8ScalarStringBytes(fragment),
          isKey,
        );
        result += fragment;
        continue;
      }
      if (isHighSurrogate(code)) {
        const low = this.text.charCodeAt(this.index + 1);
        if (!isLowSurrogate(low)) {
          this.fail("unpaired high surrogate in JSON string");
        }
        byteLength = this.checkedStringAppend(byteLength, 4, isKey);
        result += this.text.slice(this.index, this.index + 2);
        this.index += 2;
        continue;
      }
      if (isLowSurrogate(code)) {
        this.fail("unpaired low surrogate in JSON string");
      }
      byteLength = this.checkedStringAppend(
        byteLength,
        utf8CodeUnitBytes(code),
        isKey,
      );
      result += this.text[this.index] as string;
      this.index += 1;
    }
    this.fail("unterminated JSON string");
  }

  private checkedStringAppend(
    currentBytes: number,
    fragmentBytes: number,
    isKey: boolean,
  ): number {
    const nextBytes = currentBytes + fragmentBytes;
    const individualLimit = isKey ? this.limits.maxKeyBytes : this.limits.maxStringBytes;
    if (!Number.isSafeInteger(nextBytes) || nextBytes > individualLimit) {
      this.fail(`decoded string exceeds ${individualLimit} bytes`);
    }
    const nextTotal = this.totalStringBytes + nextBytes;
    if (!Number.isSafeInteger(nextTotal) || nextTotal > this.limits.maxTotalStringBytes) {
      this.fail(`total decoded string bytes exceed ${this.limits.maxTotalStringBytes}`);
    }
    return nextBytes;
  }

  private parseEscape(): string {
    const escaped = this.text[this.index];
    this.index += 1;
    switch (escaped) {
      case '"':
      case "\\":
      case "/":
        return escaped;
      case "b":
        return "\b";
      case "f":
        return "\f";
      case "n":
        return "\n";
      case "r":
        return "\r";
      case "t":
        return "\t";
      case "u": {
        const first = this.readHexCodeUnit();
        if (isHighSurrogate(first)) {
          if (this.text.slice(this.index, this.index + 2) !== "\\u") {
            this.fail("escaped high surrogate lacks a low-surrogate escape");
          }
          this.index += 2;
          const second = this.readHexCodeUnit();
          if (!isLowSurrogate(second)) {
            this.fail("escaped high surrogate has an invalid pair");
          }
          return String.fromCharCode(first, second);
        }
        if (isLowSurrogate(first)) {
          this.fail("unpaired escaped low surrogate in JSON string");
        }
        return String.fromCharCode(first);
      }
      default:
        this.fail("invalid JSON string escape");
    }
  }

  private readHexCodeUnit(): number {
    const value = this.text.slice(this.index, this.index + 4);
    if (!/^[0-9a-fA-F]{4}$/.test(value)) {
      this.fail("invalid Unicode escape in JSON string");
    }
    this.index += 4;
    return Number.parseInt(value, 16);
  }

  private parseInteger(): number {
    const start = this.index;
    this.consume("-");
    if (this.consume("0")) {
      if (isDigit(this.text[this.index])) {
        this.fail("integer has a leading zero");
      }
    } else {
      const first = this.text[this.index];
      if (first === undefined || first < "1" || first > "9") {
        this.fail("invalid integer");
      }
      this.index += 1;
      while (isDigit(this.text[this.index])) {
        this.index += 1;
      }
    }
    if ([".", "e", "E"].includes(this.text[this.index] ?? "")) {
      this.fail("floating-point JSON numbers are forbidden");
    }
    const token = this.text.slice(start, this.index);
    if (token === "-0") {
      this.fail("negative zero is forbidden");
    }
    if (token.length > this.limits.maxIntegerCharacters) {
      this.fail(
        `integer has more than ${this.limits.maxIntegerCharacters} characters`,
      );
    }
    const value = Number(token);
    if (!Number.isSafeInteger(value)) {
      this.fail("integer is outside the exact JSON-safe range");
    }
    return value;
  }

  private expectLiteral(expected: string): void {
    if (this.text.slice(this.index, this.index + expected.length) !== expected) {
      this.fail(`expected ${expected}`);
    }
    this.index += expected.length;
  }

  private skipWhitespace(): void {
    while ([" ", "\t", "\r", "\n"].includes(this.text[this.index] ?? "")) {
      this.index += 1;
    }
  }

  private consume(expected: string): boolean {
    if (this.text[this.index] !== expected) {
      return false;
    }
    this.index += 1;
    return true;
  }

  private fail(message: string): never {
    throw new StrictJsonError(`${message} at UTF-16 offset ${this.index}`);
  }
}

function isDigit(value: string | undefined): boolean {
  return value !== undefined && value >= "0" && value <= "9";
}

function isHighSurrogate(value: number): boolean {
  return value >= 0xd800 && value <= 0xdbff;
}

function isLowSurrogate(value: number): boolean {
  return value >= 0xdc00 && value <= 0xdfff;
}

function utf8CodeUnitBytes(value: number): number {
  if (value <= 0x7f) return 1;
  if (value <= 0x7ff) return 2;
  return 3;
}

function utf8ScalarStringBytes(value: string): number {
  if (value.length === 2) return 4;
  const code = value.charCodeAt(0);
  if (!Number.isFinite(code)) throw new StrictJsonError("empty decoded string fragment");
  return utf8CodeUnitBytes(code);
}
